src/litdata/streaming/writer.py

```diff
@@ -370,6 +374,10 @@ def write_chunk_to_file(
         if self._compression:
             raw_data = self._compressor.compress(raw_data)
 
+        # Whether to encrypt the raw bytes
+        if self._encryption:
```

Comment:
I wonder if it is better to offer to encrypt either each sample or the entire chunk. My thinking is that if you encrypt each sample, then you can have still read each sample independently.

Reply:
Oh, now I understand. There was a parameter level in the example:
```
encryption=Entryption(password="...", level="sample|chunk")
```

Reply:
I think compression should be used before encryption.

Reply:
Sure @tchaton ,
Here's a brief example I used to generate the insights. 
I also tested with some real images, and the ratios were quite similar.

So in this PR, I followed the sequence of encrypting before compressing.

####  Image Compression and Encryption Insights

| S.N. | Stage                       | Size (MB) |
|------|-----------------------------|-----------|
| 0    | Original Image              | 3.000000  |
| 1    | Compressed Image            | 3.000078  |
| 2    | Decompressed Image          | 3.000000  |
| 3    | Encrypted Image             | 4.000095  |
| 4    | Compressed After Encryption | 3.000707  |
| 5    | Encrypted After Compression | 4.000195  |

<details>
  <summary>Test Script</summary>

```python
# !pip install zstd cryptography

import zstd
import numpy as np
import pandas as pd
from PIL import Image
from cryptography.fernet import Fernet

# Generate a key and create a Fernet encryptor
key = Fernet.generate_key()
fernet = Fernet(key)

calculate_size = lambda x: len(x) / (1024 * 1024)

def generate_insights(data_bytes: bytes, encryptor: Fernet) -> pd.DataFrame:
    # Only compression
    compressed_bytes = zstd.compress(data_bytes)

    # Only encryption
    encrypted_bytes = encryptor.encrypt(data_bytes)

    # Compression after encryption
    compressed_after_encryption_bytes = zstd.compress(encrypted_bytes)

    # Encryption after compression
    encrypted_after_compression_bytes = encryptor.encrypt(compressed_bytes)

    # Decompression to get original bytes back
    decompressed_bytes = zstd.decompress(compressed_bytes)

    # Create a DataFrame to display results
    data = {
        "S.N.": [
            "Original Image",
            "Compressed Image",
            "Decompressed Image",
            "Encrypted Image",
            "Compressed After Encryption",
            "Encrypted After Compression"
        ],
        "Size (MB)": [
            calculate_size(data_bytes),
            calculate_size(compressed_bytes),
            calculate_size(decompressed_bytes),
            calculate_size(encrypted_bytes),
            calculate_size(compressed_after_encryption_bytes),
            calculate_size(encrypted_after_compression_bytes)
        ]
    }

    df = pd.DataFrame(data)
    return df

# Generate a fake image
fake_image = Image.fromarray(np.random.randint(0, 256, (1024, 1024, 3), dtype=np.uint8))
img_bytes = fake_image.tobytes()

# Call the function and display insights
insights_df = generate_insights(img_bytes, fernet)
print(insights_df)
```

</details>

Reply:
Nice, so it seems we need to compress after encrypted. But it is weird the compression isn't helping much.

Reply:
Here is an example of this image:
<img src="https://images.pexels.com/photos/1108099/pexels-photo-1108099.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2" alt="Sample Image" width="300" height="200">


| S.N. | Stage                       | Size (MB) |
|------|-----------------------------|-----------|
| 0    | Original Image              | 8.583069  |
| 1    | Compressed Image            | 3.433589  |
| 2    | Decompressed Image          | 8.583069  |
| 3    | Encrypted Image             | 11.444187 |
| 4    | Compressed After Encryption | 8.585006  |
| 5    | Encrypted After Compression | 4.578197  |


I haven't checked yet for chunk vs sample. I will look into it

Reply:
In this case, it seems to have the reverse effect. 
I will do some more tests.